Add potential evapotranspiration method based on Droogers and Allen (2002)

JavierDiezSierra opened the pull request:
<!--Please ensure the PR fulfills the following requirements! -->
<!-- If this is your first PR, make sure to add your details to the AUTHORS.rst! -->
### Pull Request Checklist:
- [X] This PR addresses an already opened issue (for bug fixes / features)
  - This PR closes #1710 
- [X] Tests for the changes have been added (for bug fixes / features)
  - [X] (If applicable) Documentation has been added / updated (for bug fixes / features)
- [X] CHANGES.rst has been updated (with summary of main changes)
  - [X] Link to issue (:issue:`1710`) and pull request (:pull:`1723`) has been added

### What kind of change does this PR introduce?

This change introduces a new method to calculate potential evapotranspiration based on Droogers and Allen (2002). The method incorporates precipitation as a proxy for humidity and is more appropriate for dry regions

### Does this PR introduce a breaking change?

No

### Other information:

aulemahal reviewed (commented):
Thanks for the PR!

Please look at my commit to see if you agree. We try to avoid exact frequency matching in indices. Xclim's Indicators are the objects that may enforce strong rules about data frequency, but indices should let the user do what it wants.

So instead of deciding what to do according to the `infer_freq`, I used the same idea as in the TW48 implementation : resample to monthly no matter what. If the data is already monthly, this will be a no-op. I also generalized the function to get a daily time coord from monthly data, as used by TW48 also. 

One caveat of my technique is the chunksizes that are not easily inferable. I took a guess that passing the same chunks as monthly tasmin would be ok in most cases. Especially because we resample the radiation right away.

Would you be able to add a test case for this new method ? The test should be a method of  `class TestPotentialEvapotranspiration`,  line 3159 of `tests/test_indices.py`.

Zeitsperre commented on CHANGES.rst:
```diff
@@ -4,12 +4,16 @@ Changelog
 
+New features and enhancements
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+* Indicator ``xclim.indices.potential_evapotranspiration`` now accepts a new value (`DA02`) for argument `method` implementing potential evapotranspiration based on Droogers and Allen (2002). (:issue:`1710`, :pull:`1723`).
```
```suggestion
* Indicator ``xclim.atmos.potential_evapotranspiration`` and indice ``xclim.indices.potential_evapotranspiration`` now accept a new value (`DA02`) for argument `method` implementing potential evapotranspiration based on Droogers and Allen (2002). (:issue:`1710`, :pull:`1723`).
```

Zeitsperre commented:
@aulemahal I think that I have the workflows configured to trigger on any comment. This might be why the builds often cancel during reviews. Will open an issue.

Zeitsperre commented on docs/references.bib:
```diff
@@ -2114,3 +2114,19 @@ @article{brode_utci_2012
+	pages = {33 – 45},
+	doi = {10.1023/A:1015508322413},
+	url = {https://www.scopus.com/inward/record.uri?eid=2-s2.0-0036464359&doi=10.1023%2fA%3a1015508322413&partnerID=40&md5=7322aaa4c6874878f5b1dab3c73c1718},
+	type = {Article},
+	publication_stage = {Final},
+	source = {Scopus},
+	note = {Cited by: 747}
```
```suggestion
	type = {Article}
```

JavierDiezSierra commented:
> Thanks for the PR!
> 
> Please look at my commit to see if you agree. We try to avoid exact frequency matching in indices. Xclim's Indicators are the objects that may enforce strong rules about data frequency, but indices should let the user do what it wants.
> 
> So instead of deciding what to do according to the `infer_freq`, I used the same idea as in the TW48 implementation : resample to monthly no matter what. If the data is already monthly, this will be a no-op. I also generalized the function to get a daily time coord from monthly data, as used by TW48 also.
> 
> One caveat of my technique is the chunksizes that are not easily inferable. I took a guess that passing the same chunks as monthly tasmin would be ok in most cases. Especially because we resample the radiation right away.
> 
> Would you be able to add a test case for this new method ? The test should be a method of `class TestPotentialEvapotranspiration`, line 3159 of `tests/test_indices.py`.

Thank you for the proposed changes. It seems correct to resample everything to monthly and work at that resolution. Regarding the issue of chunksize when calculating solar radiation, it was causing problems for me, so I have removed it. On the other hand, I have modified the term lv=2.5 and included the constant 0.408 to align the Hargreaves (HG85) method with the new modified method (DA02). The result of applying 1/2.5 = 0.4 is very similar to multiplying by 0.408, but this way we use the original constant that appears in Hargreaves et al (2002).

Finally, I have added the test for the new method, although when running all tests with `pytest`, I encountered many errors in many other tests, although perhaps it's my problem and I haven't followed the correct procedure.

Zeitsperre commented:
@JavierDiezSierra This is looking good!

Once the failing tests are addressed, we should be able to move forward with the merge in advance of the next release (v0.49.0).

All the best!

aulemahal commented:
@JavierDiezSierra I think I fixed the tests in the last commit. This looks good to go.

JavierDiezSierra commented:
> @JavierDiezSierra I think I fixed the tests in the last commit. This looks good to go.

¡Many thanks for you help!.

Zeitsperre commented:
@JavierDiezSierra

Thanks so much for your pull request!
